File: jsonld-terse.js
// Copyright © 2024 Michael Thornburgh
// SPDX-License-Identifier: MIT

class com_zenomt_JSONLD_Terse {
	constructor(node, { documentUri, vocab, fallbackContext, maxDepth=64 } = {}) {
		this._nodes = new Map();
		this._literals = new Map();
		this._root = this.get(node ? this.merge(node, { documentUri, vocab, maxDepth, fallbackContext }) : null) ?? this._nodes.values().next().value;
	}

	merge(node, { documentUri, vocab, fallbackContext, maxDepth=64 } = {}) {
		let baseUri, prefixes;
		({ baseUri, prefixes, vocab } = this.constructor._resolveContext(documentUri, vocab, fallbackContext, {}));
		return this._basicMerge(node, { depth: 0, visited: new Map(), blankNodes: new Map(), prefixes, baseUri, vocab, maxDepth, literals: this._literals });
	}

	get(uriOrNode) { return this._nodes.get(uriOrNode) ?? (uriOrNode ? this._nodes.get(uriOrNode["@id"]) : null); }

	get nodes() { return new Set(this._nodes.values()) ; }

	get root() { return this._root; }
	set root(uriOrNode) { this._root = this.get(uriOrNode); }

	asTriples() {
		let nextBlank = 0;
		const blanks = new Map();
		const getId = node => node["@id"] ?? blanks.get(node) ?? blanks.set(node, "_:b" + nextBlank++).get(node);
		const valueObject = node => {
			if(Array.isArray(node))
				return node.map(valueObject);
			if("@list" in node)
				return { "@list": node["@list"].map(valueObject) };
			if("@value" in node)
				return this.constructor._adaptLiteral(node);
			return { "@id": getId(node) };
		}

		const rv = [];
		for(const node of this.nodes)
		{
			const subject = getId(node);
			for(const [predicate, values] of Object.entries(node))
				if(predicate[0] != "@")
					for(const value of values)
						rv.push({ subject, predicate, _object: valueObject(value) });
		}

		return rv;
	}

	asTree(root, { noArray, base, rawLiteral } = {}) {
		root = root ? this.get(root) : this.root;
		base = base ? (new URL("", base)) : null;
		const basePath = base ? (new URL(".", base)) : null;
		const baseRoot = base ? (new URL("/", base)) : null;
		const context = { visited: new Map(), nextBlank: 0, noArray, base, basePath, baseRoot, rawLiteral };

		const rv =  root ? this._basicAsTree(root, context) : {};

		const included = [];
		for(const node of this.nodes)
			if(!context.visited.has(node))
				included.push(this._basicAsTree(node, context));
		if(included.length > 0)
			rv["@included"] = included;

		return rv;
	}

	asJSON(root, { noArray, indent, base, rawLiteral } = {}) { return JSON.stringify(this.asTree(root, { noArray, base, rawLiteral }), null, indent); }

	static effectiveRootContext(node, { documentUri, vocab, fallbackContext } = {}) {
		let baseUri, prefixes;
		({ baseUri, prefixes, vocab } = this._resolveContext(documentUri, vocab, fallbackContext, {}));
		({ baseUri, prefixes, vocab } = this._resolveContext(baseUri, vocab, node?.["@context"], prefixes));
		prefixes["@base"] = baseUri?.href ?? baseUri;
		prefixes["@vocab"] = vocab;
		return prefixes;
	}

	_makeRelative(uri, { base, basePath, baseRoot }) {
		if((!base) || (base.origin != uri.origin))
			return uri.href;
		if((new URL("", uri)).href == base.href)
			return uri.hash;
		if((new URL(".", uri)).href.startsWith(basePath.href))
			return uri.href.substring(basePath.href.length) || ".";
		return uri.href.substring(baseRoot.href.length - 1);
	}

	_basicAsTree(node, context) {
		if(Array.isArray(node))
			return ((node.length == 1) && context.noArray) ? this._basicAsTree(node[0], context) : node.map(v => this._basicAsTree(v, context));
		if("@list" in node)
			return { "@list": node["@list"].map(v => this._basicAsTree(v, context)) };
		if("@value" in node)
			return this.constructor._adaptLiteral(node, { rawLiteral: context.rawLiteral } );

		const item = context.visited.get(node);
		if(item)
		{
			if(!("@id" in item))
				item["@id"] = "_:b" + context.nextBlank++;
			return { "@id": item["@id"] };
		}

		const rv = {};
		context.visited.set(node, rv);

		for(const [key, value] of Object.entries(node))
			rv[key] = "@id" == key ? this._makeRelative(new URL(value), context) : this._basicAsTree(value, context);

		return rv;
	}

	_basicMerge(node, { depth, visited, blankNodes, prefixes, baseUri, vocab, maxDepth, literals }) {
		if(++depth > maxDepth)
			throw new RangeError("nested too deep");

		({ baseUri, prefixes, vocab } = this.constructor._resolveContext(baseUri, vocab, node?.["@context"], prefixes));
		const context = { depth, visited, blankNodes, prefixes, baseUri, vocab, maxDepth, literals };

		if(Array.isArray(node))
			return node.map(v => this._basicMerge(v, context));
		if(this.constructor._isPrimitive(node))
			return this.constructor._adaptLiteral({ "@value": node ?? null }, context);
		if("@list" in node)
			return { "@list": Array.isArray(node["@list"]) ? node["@list"].map(v => this._basicMerge(v, context)) : [] };
		if("@value" in node)
			return this.constructor._adaptLiteral(node, context);
		if(visited.has(node))
			return visited.get(node);

		const uri = this.constructor._expandUri(node["@id"], false, context);
		const isBlank = (typeof(uri) != "string") || uri.substring(0, 2) == "_:";
		const rv = (isBlank ? blankNodes.get(uri) : this._nodes.get(uri)) ?? {};
		if(!isBlank)
			rv["@id"] = uri;
		else if(uri)
			blankNodes.set(uri, rv);
		this._nodes.set(isBlank ? rv : uri, rv);
		visited.set(node, rv);

		for(let [key, value] of Object.entries(node))
		{
			if("@type" == key)
			{
				key = "http://www.w3.org/1999/02/22-rdf-syntax-ns#type";
				value = (Array.isArray(value) ? value : [value]).map(v => ({ "@id": this.constructor._expandUri(v, !!context.vocab, context) }));
			}

			if("@included" == key)
				this._basicMerge(value, context);
			else if(key[0] != "@")
			{
				key = this.constructor._expandUri(key, true, context);
				if(key)
					rv[key] = Array.from(new Set((rv[key] ?? []).concat((Array.isArray(value) ? value : [value]).map(v => this._basicMerge(v, context)))));
			}
		}

		return rv;
	}

	static _resolveUri(uri, baseUri) { return uri ? (new URL(uri, baseUri)) : baseUri; }

	static _resolveVocab(context, vocab, baseUri) {
		vocab = context?.["@vocab"] ?? vocab;
		return (vocab != null) ? (new URL(vocab, baseUri)).href : null;
	}

	static _overlayPrefixes(base, overlay, baseUri) {
		if(overlay == null)
			return base;
		const rv = {};
		for(const [key, value] of Object.entries(base))
			rv[key] = value;
		for(const [key, value] of Object.entries(overlay ?? {}))
			if(key[0] != "@")
				rv[key] = (typeof(value) == "string") ? (new URL(value, baseUri)).href : null;
		return rv;
	}

	static _expandUri(uri, isKey, { prefixes, vocab, baseUri }) {
		if(typeof(uri) != "string")
			return null;
		if(uri == "@json")
			return "http://www.w3.org/1999/02/22-rdf-syntax-ns#JSON";
		const colonPosition = uri.indexOf(":");
		if(colonPosition >= 0)
		{
			const prefix = uri.substring(0, colonPosition);
			return (typeof(prefixes[prefix]) == "string") ? prefixes[prefix] + uri.substring(colonPosition + 1) : uri;
		}
		return prefixes[uri] ?? (isKey ? (vocab ? vocab + uri : null) : (new URL(uri, baseUri)).href);
	}

	static _adaptLiteral(node, { prefixes = {}, baseUri, literals, rawLiteral } = {}) {
		if(rawLiteral && ["@type", "@language", "@direction"].every(key => !(key in node)) && this._isPrimitive(node["@value"]))
			return structuredClone(node["@value"]);
		const rv = {};
		[ "@value", "@language", "@direction" ].forEach(key => { if(key in node) rv[key] = structuredClone(node[key]) });
		if(typeof(node["@type"]) == "string")
			rv["@type"] = this._expandUri(node["@type"], false, { prefixes, baseUri });
		const valueKey = JSON.stringify([rv["@value"], rv["@type"], rv["@language"]]);
		return literals ? literals.get(valueKey) ?? (literals.set(valueKey, rv), rv) : rv;
	}

	static _isPrimitive(value) { return (typeof(value) != "object") || (null == value); }

	static _resolveContext(baseUri, vocab, context, prefixes) {
		baseUri = this._resolveUri(context?.["@base"], baseUri);
		prefixes = this._overlayPrefixes(prefixes, context, baseUri);
		vocab = this._resolveVocab(context, vocab, baseUri);
		return { baseUri, prefixes, vocab };
	}
}

File: jsonldterse.py
# Copyright © 2025 Michael Thornburgh
# SPDX-License-Identifier: MIT

from urllib.parse import urlsplit, urljoin, urldefrag
import copy
import json

class JSONLD_Terse:
	def __init__(self, node = None, documentUri = None, vocab = None, fallbackContext = None, maxDepth = 64):
		self._nodes = {}
		self._uris = {}
		self._literals = {}
		self._root = self.get(self.merge(node, documentUri=documentUri, vocab=vocab, maxDepth=maxDepth, fallbackContext=fallbackContext) if node else None) or self._first(self._nodes.values())

	def merge(self, node, documentUri=None, vocab=None, fallbackContext=None, maxDepth=64):
		baseUri, prefixes, vocab = self._resolveContext(documentUri, vocab, fallbackContext, {})
		return self._basicMerge(node, depth=0, visited={}, blankNodes={}, prefixes=prefixes, baseUri=baseUri, vocab=vocab, maxDepth=maxDepth, literals=self._literals)

	def get(self, uriOrNode):
		if type(uriOrNode) == str:
			return self._uris.get(uriOrNode, None)
		if type(uriOrNode) == dict:
			if id(uriOrNode) in self._nodes:
				return uriOrNode
			return self._uris.get(uriOrNode.get("@id", None), None)

	@property
	def nodes(self):
		return list(self._nodes.values())

	@property
	def root(self):
		return self._root or self._first(self._nodes.values())

	@root.setter
	def root(self, uriOrNode):
		self._root = self.get(uriOrNode)

	def asTriples(self):
		nextBlank = 0
		blanks = {}
		def getId(node):
			nonlocal nextBlank
			rv = node.get("@id", None) or blanks.get(id(node), None)
			if rv != None:
				return rv
			name = "_:b" + str(nextBlank)
			nextBlank += 1
			blanks[id(node)] = name
			return name
		def valueObject(node):
			if type(node) == list:
				return list(map(valueObject, node))
			if "@list" in node:
				return { "@list": list(map(valueObject, node["@list"])) }
			if "@value" in node:
				return self._adaptLiteral(node)
			return { "@id": getId(node) }
		rv = []
		for node in self.nodes:
			subject = getId(node)
			for predicate, values in node.items():
				if predicate[:1] != "@":
					for value in values:
						rv.append(dict(subject=subject, predicate=predicate, _object=valueObject(value)))
		return rv

	def asTree(self, root = None, noArray = False, base = None, rawLiteral = False):
		root = self.get(root) if root else self.root
		base = self._resolveUri("", base) if base is not None else None
		basePath = self._resolveUri(".", base) if base is not None else None
		baseRoot = self._resolveUri("/", base) if base is not None else None
		ctx = dict(visited={}, nextBlank=0, noArray=noArray, base=base, basePath=basePath, baseRoot=baseRoot, rawLiteral=rawLiteral)
		rv = self._basicAsTree(root, ctx) if root is not None else {}
		included = []
		for ident, node in self._nodes.items():
			if (ident not in ctx["visited"]) and any(map(lambda k: k[:1] != "@", node.keys())):
				included.append(self._basicAsTree(node, ctx))
		if len(included):
			rv["@included"] = included
		return rv

	def asJSON(self, root = None, indent = None, separators = None, noArray = False, base = None, rawLiteral = False):
		return json.dumps(self.asTree(root, noArray=noArray, base=base, rawLiteral=rawLiteral), indent=indent, separators=separators)

	def select(self, s = None, p = None, o = None, literal = None, nodes = None, column = None, filter = None):
		rv = []
		querySubjectNode = self.get(s)
		queryPredicateNode = self.get(p)
		queryObjectNode = self.get(o)
		nodes = nodes if nodes is not None else ([querySubjectNode] if s is not None else self._nodes.values())
		literal = ({ "@value": literal } if self._isPrimitive(literal) else literal) if literal is not None else None
		if (s is not None and querySubjectNode is None) or (p is not None and queryPredicateNode is None) or (o is not None and queryObjectNode is None):
			return rv
		queryPredicateUri = queryPredicateNode.get("@id", None) if queryPredicateNode is not None else None
		for subject in nodes:
			if querySubjectNode is not None and subject is not querySubjectNode:
				continue
			for key, values in (([(queryPredicateUri, subject.get(queryPredicateUri, []))]) if p is not None else subject.items()):
				if key[:1] != "@":
					predicate = queryPredicateNode or self.get(key)
					for _object in values:
						if queryObjectNode is not None and _object != queryObjectNode:
							continue
						if literal is not None:
							if ("@value" not in _object) or any(map(lambda k: (k in literal) and literal[k] != _object.get(k, None), ["@value", "@language", "@direction", "@type"])):
								continue
						if filter is not None and not filter(subject, predicate, _object):
							continue
						rv.append(dict(subject=subject, predicate=predicate, _object=_object))
		return self._unique(map(lambda each: each.get(column, None), rv)) if column else rv

	@classmethod
	def effectiveRootContext(cls, node, documentUri = None, vocab = None, fallbackContext = None):
		baseUri, prefixes, vocab = cls._resolveContext(documentUri, vocab, fallbackContext, {})
		baseUri, prefixes, vocab = cls._resolveContext(baseUri, vocab, node.get("@context", None) if node else None, prefixes)
		if baseUri:
			prefixes["@base"] = baseUri
		if vocab:
			prefixes["@vocab"] = vocab
		return prefixes

	def _makeRelative(self, uri, base, basePath, baseRoot, **unused):
		if not base:
			return uri
		uri_split = urlsplit(uri)
		base_split = urlsplit(base)
		if uri_split.scheme != base_split.scheme or uri_split.netloc != base_split.netloc:
			return uri
		if urldefrag(uri)[0] == base:
			return ("#" if uri_split.fragment or uri[-1:] == "#" else "") + uri_split.fragment
		if self._resolveUri(".", uri).startswith(basePath):
			return uri[len(basePath):] or "."
		return uri[len(baseRoot) - 1:]

	def _basicAsTree(self, node, ctx):
		if type(node) == list:
			return self._basicAsTree(node[0], ctx) if len(node) == 1 and ctx["noArray"] else list(map(lambda v: self._basicAsTree(v, ctx), node))
		if "@list" in node:
			return { "@list": list(map(lambda v: self._basicAsTree(v, ctx), node["@list"])) }
		if "@value" in node:
			return self._adaptLiteral(node, rawLiteral=ctx["rawLiteral"])
		item = ctx["visited"].get(id(node))
		if item is not None:
			if "@id" not in item:
				item["@id"] = "_:b" + str(ctx["nextBlank"])
				ctx["nextBlank"] += 1
			return { "@id": item["@id"] }
		rv = {}
		ctx["visited"][id(node)] = rv
		for key, value in node.items():
			rv[key] = self._makeRelative(value, **ctx) if key == "@id" else self._basicAsTree(value, ctx)
		return rv

	def _basicMerge(self, node, depth, visited, blankNodes, prefixes, baseUri, vocab, maxDepth, literals):
		depth += 1
		if depth > maxDepth:
			raise RecursionError("nested too deep")

		baseUri, prefixes, vocab = self._resolveContext(baseUri, vocab, node.get("@context", None) if type(node) == dict else None, prefixes)
		ctx = dict(depth=depth, visited=visited, blankNodes=blankNodes, prefixes=prefixes, baseUri=baseUri, vocab=vocab, maxDepth=maxDepth, literals=literals)

		if type(node) == list:
			return list(map(lambda v: self._basicMerge(v, **ctx), node))
		if self._isPrimitive(node):
			return self._adaptLiteral({"@value": node}, **ctx)
		if "@list" in node:
			return { "@list": list(map(lambda v: self._basicMerge(v, **ctx), node["@list"])) if type(node["@list"]) == list else [] }
		if "@value" in node:
			return self._adaptLiteral(node, **ctx)
		if id(node) in visited:
			return visited[id(node)]

		uri = self._expandUri(node.get("@id", None), False, **ctx)
		isBlank = (type(uri) != str) or uri[:2] == "_:"
		rv = blankNodes.get(uri, {}) if isBlank else self._uris.get(uri, {})
		if not isBlank:
			rv["@id"] = uri
		elif uri:
			blankNodes[uri] = rv
		self._nodes[id(rv)] = rv
		if not isBlank:
			self._uris[uri] = rv
		visited[id(node)] = rv

		for key, value in node.items():
			if key == "@type":
				key = "http://www.w3.org/1999/02/22-rdf-syntax-ns#type"
				value = list(map(lambda v: { "@id": self._expandUri(v, isKey = bool(vocab), **ctx) }, value if type(value) == list else [value]))

			if key == "@included":
				self._basicMerge(value, **ctx)
			elif key[:1] != "@":
				key = self._expandUri(key, True, **ctx)
				if key:
					if key not in self._uris:
						keyNode = { "@id": key }
						self._uris[key] = keyNode
						self._nodes[id(keyNode)] = keyNode
					rv[key] = self._unique(rv.get(key, []) + list(map(lambda v: self._basicMerge(v, **ctx), value if type(value) == list else [value])))

		return rv

	@staticmethod
	def _resolveUri(uri, baseUri):
		if uri is not None and not baseUri:
			return uri
		if uri is not None:
			rv = urldefrag(baseUri)[0] if uri == '' else urljoin(baseUri, uri)
			if uri[-1:] == "#" and rv[-1:] != "#":
				# bug in urljoin, empty fragment is distinct from no fragment!
				rv += "#"
			return rv
		return baseUri

	@classmethod
	def _resolveVocab(cls, context, vocab, baseUri):
		vocab = context.get("@vocab", vocab) if context else vocab
		return cls._resolveUri(vocab, baseUri) if vocab != None else None

	@classmethod
	def _overlayPrefixes(cls, base, overlay, baseUri):
		if overlay == None:
			return base
		rv = dict(base)
		for key, value in (overlay or {}).items():
			if key[:1] != "@":
				rv[key] = cls._resolveUri(value, baseUri) if type(value) == str else None
		return rv

	@classmethod
	def _expandUri(cls, uri, isKey = False, prefixes = {}, vocab = None, baseUri = None, **unused):
		if type(uri) != str:
			return None
		if uri == "@json":
			return "http://www.w3.org/1999/02/22-rdf-syntax-ns#JSON"
		colonPosition = uri.find(":")
		if colonPosition >= 0:
			prefix = prefixes.get(uri[:colonPosition], None)
			return prefix + uri[colonPosition + 1:] if type(prefix) == str and uri.find("://") != colonPosition else uri
		return prefix if (prefix := prefixes.get(uri, None)) else ((vocab + uri if vocab else None) if isKey else cls._resolveUri(uri, baseUri))

	@classmethod
	def _adaptLiteral(cls, node, prefixes = None, baseUri = None, literals = None, rawLiteral = False, **unused):
		prefixes = prefixes or {}
		if rawLiteral and all(map(lambda key : key not in node, ["@type", "@language", "@direction"])) and cls._isPrimitive(node["@value"]):
			return copy.deepcopy(node["@value"])
		rv = {}
		for key in ["@value", "@language", "@direction"]:
			if key in node:
				rv[key] = copy.deepcopy(node[key])
		if type(node.get("@type", None)) == str:
			rv["@type"] = cls._expandUri(node["@type"], False, prefixes=prefixes, baseUri=baseUri)
		if literals is not None:
			valueKey = json.dumps([rv["@value"], rv.get("@type", None), rv.get("@language", None)])
			if valueKey in literals:
				return literals.get(valueKey)
			literals[valueKey] = rv
		return rv

	@staticmethod
	def _isPrimitive(value):
		return type(value) not in [dict, list]

	@classmethod
	def _resolveContext(cls, baseUri, vocab, context, prefixes):
		baseUri = cls._resolveUri(context.get("@base", None) if context else None, baseUri)
		prefixes = cls._overlayPrefixes(prefixes, context, baseUri)
		vocab = cls._resolveVocab(context, vocab, baseUri)
		return baseUri, prefixes, vocab

	@staticmethod
	def _unique(iterable):
		seen = set()
		rv = []
		for each in iterable:
			if id(each) not in seen:
				seen.add(id(each))
				rv.append(each)
		return rv

	@staticmethod
	def _first(iterable):
		for i in iterable:
			return i
